components/event-bus/internal/knutil/eventing.go

```diff
@@ -0,0 +1,270 @@
+			return err
+		}
+	} else if res.StatusCode != http.StatusAccepted {
+		log.Printf("ERROR: SendMessage(): %s", res.Status)
+		return errors.New(res.Status)
+	}
+	// ok
+	return nil
+}
+
+func resendMessage(httpClient *http.Client, channel *evapisv1alpha1.Channel, message *string)  error {
+	timeout := time.After(10 * time.Second)
+	tick := time.Tick(200 * time.Millisecond)
+	req, err := makeHttpRequest(channel, message)
+	if err != nil {
+		log.Printf("ERROR: resendMessage(): makeHttpRequest() failed: %v", err)
+		return err
+	}
+	res, err := httpClient.Do(req)
+	if err != nil {
+		log.Printf("ERROR: resendMessage(): could not send HTTP request: %v", err)
+		return err
+	}
+	defer res.Body.Close()
+	//dumpResponse(res)
+	sc := res.StatusCode
+	for ; sc == http.StatusNotFound; {
+		select {
+		case <-timeout:
+			log.Printf("ERROR: resendMessage(): timed out")
+			return errors.New("ERROR: timed out")
+		case <-tick:
+			req, err := makeHttpRequest(channel, message)
+			if err != nil {
+				log.Printf("ERROR: resendMessage(): makeHttpRequest() failed: %v", err)
+				return err
+			}
+			res, err := httpClient.Do(req)
+			if err != nil {
+				log.Printf("ERROR: resendMessage(): could not resend HTTP request: %v", err)
+				return err
+			}
+			defer res.Body.Close()
+			dumpResponse(res)
+			sc = res.StatusCode
+		}
+	}
+	if sc != http.StatusAccepted {
+		log.Printf("ERROR: resendMessage(): %v", sc)
+		return errors.New(string(sc))
+	}
+	return nil
+}
+
+func createChannel(evClient eventingv1alpha1.EventingV1alpha1Interface, provisioner string, name string, namespace string, timeout time.Duration) (*evapisv1alpha1.Channel, error) {
+	c := makeChannel(provisioner, name, namespace)
+	if channel, err := evClient.Channels(namespace).Create(c); err != nil && !k8serrors.IsAlreadyExists(err) {
+		log.Printf("ERROR: createChannel(): creating channel: %v", err)
+		return nil, err
+	} else {
+		isReady := channel.Status.IsReady()
+		tout := time.After(timeout)
+		tick := time.Tick(100 * time.Millisecond)
+		for ; !isReady; {
+			select {
+			case <-tout:
+				return nil, errors.New("timed out")
+			case <-tick:
+				if channel, err = evClient.Channels(namespace).Get(name, metav1.GetOptions{}); err != nil {
+					log.Printf("ERROR: createChannel(): geting channel: %v", err)
+				} else {
+					isReady = channel.Status.IsReady()
+				}
+			}
+		}
+		return channel, nil
+	}
+}
+
+func makeChannel(provisioner string, name string, namespace string) *evapisv1alpha1.Channel {
+	c := &evapisv1alpha1.Channel{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: evapisv1alpha1.SchemeGroupVersion.String(),
+			Kind:       "Channel",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      name,
+		},
+		Spec: evapisv1alpha1.ChannelSpec{
+			Provisioner: &corev1.ObjectReference{
+				Name: provisioner,
+			},
+		},
+	}
+	return c
+}
+
+func makeHttpRequest(channel *evapisv1alpha1.Channel, message *string) (*http.Request, error) {
```

Comment:
should we consider decoupling the code that is doing the HTTP work or other such low-level tasks from the KN interaction?

Reply:
It could evolve in the future, now is very simple.